source/common/http/conn_manager_impl.cc

```diff
@@ -611,12 +612,23 @@ ConnectionManagerImpl::ActiveStream::~ActiveStream() {
 
   if (active_span_) {
     Tracing::HttpTracerUtility::finalizeDownstreamSpan(
-        *active_span_, filter_manager_.requestHeaders(), filter_manager_.responseHeaders(),
-        filter_manager_.responseTrailers(), filter_manager_.streamInfo(), *this);
+        *active_span_, request_headers_.get(), response_headers_.get(), response_trailers_.get(),
+        filter_manager_.streamInfo(), *this);
   }
   if (state_.successful_upgrade_) {
     connection_manager_.stats_.named_.downstream_cx_upgrades_active_.dec();
   }
+
+  // The headers might still be owned by the FM in case we failed during filter processing (e.g. due
+  // to exceeding the buffer limits).
+  auto* response_headers = filter_manager_.responseHeaders() ? filter_manager_.responseHeaders()
```

Comment:
yeah, between this and the pass back and forth logic I feel like we'd be better off with either non-explicit ownership (much as I hate shared/weak pointers) or a different model (yet as discussed on slack I am not sure what that would be).

The pass back and forth leaves me super worried about lifetime and safety issues.

For downstream I like headers being owned by the HCM/downstreamFM unit, as it outlives all filters and makes that "easy"

We can't leave ownership permanently in the FM based class, because we'll have upstreamFM and downstreamFM and they can't both own, so if ownership is in the FM we have to hand off at some point.

That makes me want to move downstream ownership to HCM, and upstream ownership to [something - maybe upstreamFM subclass of baseFM?] so that the baseFM never owns and we can avoid handoff for the whole downstream case, all FMs can assert that header references outlive FM, and only the upstreamFM has to deal with the handoff problem.

It still feels like there ought to be a better solution than that though - I'm going to escalate to level 2 Envoy data plane support and summon @mattklein123  ;-)

Reply:
![image](https://user-images.githubusercontent.com/2023366/91451194-2b9b2300-e84b-11ea-8d62-a58816fa910b.png)

I made a drawing! 

The challenge is making sure that we can express the correct ownership at both levels, given that the upstream FM cannot retain the headers for its lifetime. This PR goes the direction of loosening the ownership invariants s.t. it can be used in either location, but the other option is to make per use case variations (via sub classing/templating) to express different ownership models.

Reply:
> That makes me want to move downstream ownership to HCM, and upstream ownership to [something - maybe upstreamFM subclass of baseFM?] so that the baseFM never owns and we can avoid handoff for the whole downstream case, all FMs can assert that header references outlive FM, and only the upstreamFM has to deal with the handoff problem.

I haven't thought about it a ton, but my initial take on this PR is that we should just have downstream/upstream headers always owned by the HCM. Can we just have the downstream/upstream FM access/replace the headers via callbacks? I think this would end up being a lot simpler to understand?

Reply:
I think that would work, as long as we make sure that we don't actually keep headers/trailers around from a previous attempt (not possible to retry after trailers today, but maybe in the future). 

I'll give that a go in a separate PR to see what that would look like